tests/cxx/template_args.cc

```diff
@@ -94,6 +98,23 @@ void NestedTemplateArguments() {
   // IWYU: IndirectClass needs a declaration
   Outer<Inner<IndirectClass> >* opi;
   (void)opi;
+
+  // Test that use of template specialization type template argument is not
+  // hidden by any sugar in the AST.
+
+  // IWYU: IndirectClass is...*indirect.h
+  Outer<decltype(StaticTemplateFieldStruct::tpl)> osi;
```

Comment:
Is `decltype` only used to introduce sugar here? I wonder if there's a lighter-weight mechanism we could use instead, like a plain typedef. I'm still unused to `decltype`, but I see it's been around since C++11, so maybe it's well-established :).

It just makes me a little uncomfortable that this is the only place it's used in IWYU's test suite, so if IWYU handling of `decltype` fails, e.g. due to some AST change in Clang, this is the only place we're going to notice, and it will be confusing.

Reply:
Unrelated note: really nice testcase showing the subtle differences between `Outer<T>`, `Outer<T*>` and `Outer<T>*`!

Reply:
> I wonder if there's a lighter-weight mechanism we could use instead, like a plain typedef.

I don't want to test it with typedef, because they have a special treatment in IWYU (they may provide underlying type or not).
>It just makes me a little uncomfortable that this is the only place it's used in IWYU's test suite, so if IWYU handling of decltype fails, e.g. due to some AST change in Clang, this is the only place we're going to notice, and it will be confusing.

I intend to commit a test with `decltype` some day:
https://github.com/bolshakov-a/include-what-you-use/commit/8ccd3ed102e0f86102efb1b17ac908aff79b0099#diff-91d0eae50131503e6661637be7ec19becc0ff3d65b0a3c88f5548776e9a63396R25

Reply:
Makes sense. Let's try to get `decltype` tests in place sooner rather than later.

Reply:
BTW, I've just found out that `decltype(auto)` and `decltype(expression)` are different clang types of nodes. The former is `clang::AutoType` and the latter is `clang::DecltypeType`. Hence, a test which checks `decltype(auto)` doesn't check `decltype(expr)`, and vice versa.

Reply:
Good catch, that discovery will come in handy.